### ISSUE TITLE
memory ram usage--help! urgent

### ISSUE THREAD

**Issue by alevillard** [vlfeat/matconvnet#562]
Hi to all, in the finetuning process (and maybe also training,..I didn't try) I have a problem of too much memory ram required, and so the server kill my job. For example, I have this code of vggD and vgg-face pretrained network, and I want to finetune with a database of faces (I use getbatchDisk to charge images from Disk). the problem I see is that after the end of an epoch, I mean, all the parameters should be replaced to be used in the next epoch. What happen is that also after an epoch, or two, the memory consumption continue to increase, until occupy an amount of 200Gb of ram.. and so all memory that I can use! But for one epoch are sufficient 15Gb of ram. So, I attach here my training code to see maybe if I have forgot some option that save the ram consumption, because is not necessary to have all the variables and parameter of every epoch saved in ram. I hope that someone could reply in few time to solve this. Thank you

training code:

function [net, info] = vggD_train(imdb, expDir, varargin)

	vl_setupnn;

	% some common options
	opts.train.batchSize = 32;
	opts.train.numEpochs = 15 ;
	opts.train.continue = true ;
	opts.train.gpus = [1] ;
	opts.train.learningRate = [2e-3*ones(1, 5),  1e-3*ones(1, 20), 1e-4*ones(1,10)];
	opts.train.weightDecay = 5e-4;
	opts.train.momentum = 0.5; 
	opts.train.expDir = expDir;
	opts.train.numSubBatches = 1;
	% getBatch options
	bopts.useGpu = numel(opts.train.gpus) >  0 ;
	opts.optMethod = 'gradient'; % ['adagrad', 'gradient']

     if(numel(varargin) > 0)
		netPre = varargin{1};
    end
% network definition!
	% MATLAB handle, passed by reference
	net = dagnn.DagNN() ;

    net.addLayer('conv1_1', dagnn.Conv('size', [3 3 3 64], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'input'}, {'conv1'},  {'conv1_1f'  'conv1_1b'});
	net.addLayer('relu1_1', dagnn.ReLU(), {'conv1'}, {'relu1'}, {});
	net.addLayer('conv1_2', dagnn.Conv('size', [3 3 64 64], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu1'}, {'conv2'},  {'conv1_2f'  'conv1_2b'});
	net.addLayer('relu1_2', dagnn.ReLU(), {'conv2'}, {'relu2'}, {});
	%net.addLayer('lrn1', dagnn.LRN('param', [5 1 0.0001/5 0.75]), {'relu1'}, {'lrn1'}, {});
	net.addLayer('pool1', dagnn.Pooling('method', 'max', 'poolSize', [2, 2], 'stride', [2 2], 'pad', [0 0 0 0]), {'relu2'}, {'pool1'}, {});
    
    net.addLayer('conv2_1', dagnn.Conv('size', [3 3 64 128], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'pool1'}, {'conv3'},  {'conv2_1f'  'conv2_1b'});
	net.addLayer('relu2_1', dagnn.ReLU(), {'conv3'}, {'relu3'}, {});
	net.addLayer('conv2_2', dagnn.Conv('size', [3 3 128 128], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu3'}, {'conv4'},  {'conv2_2f'  'conv2_2b'});
	net.addLayer('relu2_2', dagnn.ReLU(), {'conv4'}, {'relu4'}, {});
	%net.addLayer('lrn2', dagnn.LRN('param', [5 1 0.0001/5 0.75]), {'relu2'}, {'lrn2'}, {});
	net.addLayer('pool2', dagnn.Pooling('method', 'max', 'poolSize', [2, 2], 'stride', [2 2], 'pad', [0 0 0 0]), {'relu4'}, {'pool2'}, {});

    net.addLayer('conv3_1', dagnn.Conv('size', [3 3 128 256], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'pool2'}, {'conv5'},  {'conv3_1f'  'conv3_1b'});
	net.addLayer('relu3_1', dagnn.ReLU(), {'conv5'}, {'relu5'}, {});
    net.addLayer('conv3_2', dagnn.Conv('size', [3 3 256 256], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu5'}, {'conv6'},  {'conv3_2f'  'conv3_2b'});
	net.addLayer('relu3_2', dagnn.ReLU(), {'conv6'}, {'relu6'}, {});
	net.addLayer('conv3_3', dagnn.Conv('size', [3 3 256 256], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu6'}, {'conv7'},  {'conv3_3f'  'conv3_3b'});
	net.addLayer('relu3_3', dagnn.ReLU(), {'conv7'}, {'relu7'}, {});
	%net.addLayer('lrn3', dagnn.LRN('param', [5 1 0.0001/5 0.75]), {'relu2'}, {'lrn2'}, {});
	net.addLayer('pool3', dagnn.Pooling('method', 'max', 'poolSize', [2, 2], 'stride', [2 2], 'pad', [0 0 0 0]), {'relu7'}, {'pool3'}, {});
    
    net.addLayer('conv4_1', dagnn.Conv('size', [3 3 256 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'pool3'}, {'conv8'},  {'conv4_1f'  'conv4_1b'});
	net.addLayer('relu4_1', dagnn.ReLU(), {'conv8'}, {'relu8'}, {});
	net.addLayer('conv4_2', dagnn.Conv('size', [3 3 512 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu8'}, {'conv9'},  {'conv4_2f'  'conv4_2b'});
	net.addLayer('relu4_2', dagnn.ReLU(), {'conv9'}, {'relu9'}, {});
	net.addLayer('conv4_3', dagnn.Conv('size', [3 3 512 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu9'}, {'conv10'},  {'conv4_3f'  'conv4_3b'});
	net.addLayer('relu4_3', dagnn.ReLU(), {'conv10'}, {'relu10'}, {});
	%net.addLayer('lrn4', dagnn.LRN('param', [5 1 0.0001/5 0.75]), {'relu2'}, {'lrn2'}, {});
	net.addLayer('pool4', dagnn.Pooling('method', 'max', 'poolSize', [2, 2], 'stride', [2 2], 'pad', [0 0 0 0]), {'relu10'}, {'pool4'}, {});
  
    net.addLayer('conv5_1', dagnn.Conv('size', [3 3 512 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'pool4'}, {'conv11'},  {'conv5_1f'  'conv5_1b'});
	net.addLayer('relu5_1', dagnn.ReLU(), {'conv11'}, {'relu11'}, {});
	net.addLayer('conv5_2', dagnn.Conv('size', [3 3 512 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu11'}, {'conv12'},  {'conv5_2f'  'conv5_2b'});
	net.addLayer('relu5_2', dagnn.ReLU(), {'conv12'}, {'relu12'}, {});
	net.addLayer('conv5_3', dagnn.Conv('size', [3 3 512 512], 'hasBias', true, 'stride', [1, 1], 'pad', [1 1 1 1]), {'relu12'}, {'conv13'},  {'conv5_3f'  'conv5_3b'});
	net.addLayer('relu5_3', dagnn.ReLU(), {'conv13'}, {'relu13'}, {});
	net.addLayer('pool5', dagnn.Pooling('method', 'max', 'poolSize', [2, 2], 'stride', [2 2], 'pad', [0 0 0 0]), {'relu13'}, {'pool5'}, {});

	net.addLayer('conv6', dagnn.Conv('size', [7 7 512 4096], 'hasBias', true, 'stride', [1, 1], 'pad', [0 0 0 0]), {'pool5'}, {'conv14'},  {'conv6f'  'conv6b'});
	net.addLayer('relu6', dagnn.ReLU(), {'conv14'}, {'relu14'}, {});
	net.addLayer('drop6', dagnn.DropOut('rate', 0.5), {'relu14'}, {'drop6'}, {});

	net.addLayer('conv7', dagnn.Conv('size', [1 1 4096 4096], 'hasBias', true, 'stride', [1, 1], 'pad', [0 0 0 0]), {'drop6'}, {'conv15'},  {'conv7f'  'conv7b'});
	net.addLayer('relu7', dagnn.ReLU(), {'conv15'}, {'relu15'}, {});
	net.addLayer('drop7', dagnn.DropOut('rate', 0.5), {'relu15'}, {'drop7'}, {});

	net.addLayer('classifier', dagnn.Conv('size', [1 1 4096 263], 'hasBias', true, 'stride', [1, 1], 'pad', [0 0 0 0]), {'drop7'}, {'classifier'},  {'conv8f'  'conv8b'});
	net.addLayer('prob', dagnn.SoftMax(), {'classifier'}, {'prob'}, {});
	net.addLayer('objective', dagnn.Loss('loss', 'log'), {'prob', 'label'}, {'objective'}, {});
	net.addLayer('error', dagnn.Loss('loss', 'classerror'), {'prob','label'}, 'error') ;
	% -- end of the network

    % initialization of the weights (CRITICAL!!!!)
    if(numel(varargin) > 0)
		initNet_FineTuning(net, netPre);
        
	else
		 initNet_He(net);
	end	
   
    %train
    info = cnn_train_dag(net, imdb, @(i,b) getBatchDisk(bopts,i,b), opts.train, 'val', find(imdb.images.set == 2)) ;
end

function initNet_FineTuning(net, netPre)
	net.initParams();

	ind = 1;
	for l=1:length(net.layers)-4
		% is a convolution layer?
		if(strcmp(class(net.layers(l).block), 'dagnn.Conv'))
			f_ind = net.layers(l).paramIndexes(1);
			b_ind = net.layers(l).paramIndexes(2);
            
            net.params(f_ind).value = netPre.params(ind).value
			ind = ind + 1;
			net.params(b_ind).value = netPre.params(ind).value
			ind = ind + 1;
          if l<= length(net.layers)-4
			net.params(f_ind).learningRate = 1e-5;
			net.params(f_ind).weightDecay = 1;
			net.params(b_ind).learningRate = 1e-5;
			net.params(b_ind).weightDecay = 1;
			
          else
            net.params(f_ind).learningRate = 1;
			net.params(f_ind).weightDecay = 1;
			net.params(b_ind).learningRate = 1;
			net.params(b_ind).weightDecay = 1;   
          end
		end
	end

	for l=length(net.layers)-3:length(net.layers)
		% is a convolution layer?
		if(strcmp(class(net.layers(l).block), 'dagnn.Conv'))
			f_ind = net.layers(l).paramIndexes(1)
			b_ind = net.layers(l).paramIndexes(2)

            
			[h,w,in,out] = size(net.params(f_ind).value)
			he_gain = 1e-2*sqrt(2/(h*w*in)) % sqrt(1/fan_in)
			net.params(f_ind).value = (1/5)*he_gain*randn(size(net.params(f_ind).value), 'single');
			net.params(f_ind).learningRate = 1;
			net.params(f_ind).weightDecay = 1;

			net.params(b_ind).value = zeros(size(net.params(b_ind).value), 'single');
			net.params(b_ind).learningRate = 0.5;
			net.params(b_ind).weightDecay = 1;
			
		end
	end
end

% getBatch for IMDBs that are too big to be in RAM
function inputs = getBatchDisk(opts, imdb, batch)
	cell_images = vl_imreadjpeg(imdb.images.filenames(batch),'numThreads', 2, 'Resize', [224 224]) ;
	%[H, W, CH] = size(cell_images{1});

	images = zeros(224, 224, 3, 'single');
    %size(images)
	for i=1:numel(cell_images)
		im = cell_images{i};
        %im=imresize(im, [224 224]);
        %size(im)
		images(:,:,:, i) = im - imdb.images.data_mean;
	end

	labels = imdb.images.labels(1,batch) ;
	if opts.useGpu > 0
  		images = gpuArray(images) ;
	end

	inputs = {'input', images, 'label', labels} ;
end

function initNet_He(net)
	net.initParams();
	for l=1:length(net.layers)
		% is a convolution layer?
		if(strcmp(class(net.layers(l).block), 'dagnn.Conv'))
			f_ind = net.layers(l).paramIndexes(1);
			b_ind = net.layers(l).paramIndexes(2);

			[h,w,in,out] = size(net.params(f_ind).value);
			he_gain = 1e-2*sqrt(2/(h*w*in)); % sqrt(1/fan_in)
			net.params(f_ind).value = (1/2)*he_gain*randn(size(net.params(f_ind).value), 'single')
			net.params(f_ind).learningRate = 1;
			net.params(f_ind).weightDecay = 1;

			net.params(b_ind).value = zeros(size(net.params(b_ind).value), 'single');
			net.params(b_ind).learningRate = 0.5;
			net.params(b_ind).weightDecay = 1;
		end
	end
end

**Comment by lenck** [comment]
Hi, I would suspect that it is caused by an improper use of `vl_imreadjpeg`. Are you sure you are not calling it somewhere else with a prefetch flag?
If you prefetch different images than those which you are going to retrieve, e.g.:
```
vl_imreadjpeg({'ima.jpg'}, 'prefetch');
vl_imreadjpeg({'imb.jpg'});
```
Then the `ima.jpg` data stays in the memory until you call `vl_imreadjpeg({'ima.jpg'});` or `clear mex`.

**Comment by alevillard** [comment]
Hi @lenck 
for reading the images I only use this setting:

cell_images = vl_imreadjpeg(imdb.images.filenames(batch),'numThreads', 2, 'Resize', [224 224]);

images = zeros(224, 224, 3, 'single');
%size(images)
for i=1:numel(cell_images)
    im = cell_images{i};
    %im=imresize(im, [224 224]);
    %size(im)
    images(:,:,:, i) = im - imdb.images.data_mean;
end

**Comment by foolwood** [comment]
@lenck 
I did the wrong way like what you said.
It takes lots of hard drive space, but I don't know how to clean it up and the location of the temp file . Could you give me some help?

**Comment by alevillard** [comment]
Hi, for my problem of ram I decided to use the classical imread matlab function in a loop for reading the batch from disk, instead of using vl_imreadjpeg. In this way I solved my problem of ram consumption. My version of matconvnet was beta-18, so maybe in lastest version they improved this function, or maybe was some my mistakes of setting, I don't know. Hope my comment to be useful.

Alevillard
